dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -217,63 +181,47 @@ Block DMFileReader::readWithFilter(const IColumn::Filter & filter)
-        if (pack_res[pack_id].isUse() && (pack_id + 1 == pack_res.size() || !pack_res[pack_id + 1].isUse()))
-        {
-            Block block = read();
-            size_t rows = block.rows();
+        const auto [new_start_pack_id, new_pack_count, new_rs_result, new_read_rows] = block_info;
+        const size_t new_start_row_offset = pack_offset[new_start_pack_id];
 
-            if (size_t passed_count = countBytesInFilter(filter, offset, rows); passed_count != rows)
+        const auto offset_begin = new_start_row_offset - start_row_offset;
+        const auto offset_end = offset_begin + new_read_rows;
+
+        Block block = readImpl(block_info);
+        block_pack_res = block_pack_res && new_rs_result;
+        if (size_t passed_count = countBytesInFilter(filter, offset_begin, new_read_rows);
+            passed_count != new_read_rows)
+        {
+            std::vector<size_t> positions;
+            positions.reserve(passed_count);
+            for (size_t i = offset_begin; i < offset_end; ++i)
             {
-                std::vector<size_t> positions;
-                positions.reserve(passed_count);
-                for (size_t p = offset; p < offset + rows; ++p)
-                {
-                    if (filter[p])
-                        positions.push_back(p - offset);
-                }
-                for (size_t i = 0; i < block.columns(); ++i)
-                {
-                    columns[i]->insertDisjunctFrom(*block.getByPosition(i).column, positions);
-                }
+                if (filter[i])
+                    positions.push_back(i - offset_begin);
             }
-            else
+            for (size_t i = 0; i < block.columns(); ++i)
             {
-                for (size_t i = 0; i < block.columns(); ++i)
-                {
-                    columns[i]->insertRangeFrom(
-                        *block.getByPosition(i).column,
-                        0,
-                        block.getByPosition(i).column->size());
-                }
+                auto column = block.getByPosition(i).column;
+                columns[i]->insertDisjunctFrom(*column, positions);
             }
-            offset += rows;
         }
-        else if (!pack_res[pack_id].isUse())
+        else
         {
-            offset += pack_stats[pack_id].rows;
+            for (size_t i = 0; i < block.columns(); ++i)
+            {
+                columns[i]->insertRangeFrom(*block.getByPosition(i).column, 0, passed_count);
+            }
         }
     }
 
-    /// 5. Restore the pack_res[last_pack_id]
-
-    if (last_pack_id < pack_res.size())
-        pack_res[last_pack_id] = next_pack_id_pack_res_cp;
-
     Block res = getHeader().cloneWithColumns(std::move(columns));
     res.setStartOffset(start_row_offset);
-    res.setRSResult(rs_result);
+    res.setRSResult(block_pack_res);
+    addSkippedRows(read_rows - total_passed_count);
```

Comment:
We should also add inside `readWithFilter`? @JinheLin 
```
    if (read_tag == ReadTag::Query && block_info.rs_result.allMatch())
        scan_context->rs_dmfile_read_with_all += block_info.pack_count;
```

Reply:
fixed.